src/Microsoft.VisualStudio.Services.Agent/HostContext.cs

```diff
@@ -175,6 +169,65 @@ public HostContext(HostType hostType, string logFile = null)
             }
         }
 
+        private ILoggedSecretMasker CreateSecretMasker()
+        {
+            // When enabled, use the OSS package-provided secret masker from
+            // Microsoft.Security.Utilities.Core. Otherwise, use the legacy
+            // secret masker from VSO.
+            bool useNewSecretMasker = AgentKnobs.EnableNewSecretMasker.GetValue(this).AsBoolean();
+
+            // When enabled, use add additional regex patterns to the secret
+            // masker. For the OSS package-provided secret masker, this will use
+            // its 'PreciselyClassifiedSecurityKeys'. For the legacy secret
+            // masker, this will use 'CredScanPatterns' implemented in this
+            // repository.
+            bool useAdditionalMaskingRegexes = AgentKnobs.EnableAdditionalMaskingRegexes.GetValue(this).AsBoolean();
+
+#pragma warning disable CA2000 // Dispose objects before losing scope. False positive: LoggedSecretMasker takes ownership.
+            ISecretMasker rawSecretMasker;
+            if (useNewSecretMasker)
+            {
+                IEnumerable<RegexPattern> patterns = Array.Empty<RegexPattern>();
+ 
+                if (useAdditionalMaskingRegexes)
+                {
+                    patterns = WellKnownRegexPatterns.PreciselyClassifiedSecurityKeys;
+                }
+
+                rawSecretMasker = new OssSecretMasker(patterns);
+            }
+            else
+            {
+                rawSecretMasker = new SecretMasker();
+            }
+#pragma warning restore CA2000 // Dispose objects before losing scope
+
+            ILoggedSecretMasker secretMasker = new LoggedSecretMasker(rawSecretMasker);
+
+            secretMasker.AddValueEncoder(ValueEncoders.JsonStringEscape, $"HostContext_{WellKnownSecretAliases.JsonStringEscape}");
+            secretMasker.AddValueEncoder(ValueEncoders.UriDataEscape, $"HostContext_{WellKnownSecretAliases.UriDataEscape}");
+            secretMasker.AddValueEncoder(ValueEncoders.BackslashEscape, $"HostContext_{WellKnownSecretAliases.UriDataEscape}");
+
+            if (useNewSecretMasker)
+            {
+                secretMasker.AddRegex(AdditionalMaskingRegexes.UrlSecretPatternNonBacktracking, $"HostContext_{WellKnownSecretAliases.UrlSecretPattern}");
+            }
+            else
+            {
+                secretMasker.AddRegex(AdditionalMaskingRegexes.UrlSecretPattern, $"HostContext_{WellKnownSecretAliases.UrlSecretPattern}");
+
+                if (useAdditionalMaskingRegexes)
+                {
+                    foreach (var pattern in AdditionalMaskingRegexes.CredScanPatterns)
+                    {
+                        secretMasker.AddRegex(pattern, $"HostContext_{WellKnownSecretAliases.CredScanPatterns}");
```

Comment:
>secretMasker.AddRegex [](http://example.com/codeflow?start=24&length=21)

So here's a small difference between OSS masker and the old one, when we initialize the rules for the old masker, we provide the origin information. This logging will not occur when adding every pattern with the OSS masker. Instead of providing the enumerable patterns to the oss masker ctor, we could enumerate the `PreciselyClassifiedSecurityKeys` rules instead and add them one-by-one.

That might allow for a bit of simplification of this set-up code. `useSecretMasker` would control initializing the URL rule and an enumerable collection of pattern strings. then set them all using the logged secret masker instance.

Reply:
Yes, I was wondering about this, but it's trickier than just that because we can only add raw regex strings via the interface.

Reply:
Ah, yes. We could rip out the raw `Pattern` property but we'd lose the id, name, and detection metadata, data which is crucial for generating correlating ids.